### ISSUE TITLE
Exception event handlers can't be triggered together

### ISSUE THREAD

**Issue by Nic-Ma** [pytorch/ignite#1234]
## 🐛 Bug description
If we attach several handlers to exception raised event, there are 2 issues:
1. If don't raise the exception again after handling, the training can't stop.
2. If raise the exception again, ignite will stop training immediately and ignore all the other handlers in the list.

I think maybe we should not raise exception in handlers, and ignite raise exception when triggered all the handlers.

Thanks.

**Comment by vfdev-5** [comment]
@Nic-Ma thanks for reporting !

> I think maybe we should not raise exception in handlers, and ignite raise exception when triggered all the handlers.

Yes, I agree this would be better

**Comment by amatsukawa** [comment]
> 1. If don't raise the exception again after handling, the training can't stop.

You can run all the handlers and stop training by calling `engine.terminate` inside the exception handler, but not re-raising the exception. I believe it is true that if you do reraise the exception from the handler, that will short circuit remaining exception handlers from running. 

You can do this: follow the convention of all fo your "meaningful" exception handlers doing whatever they need to do, but *not* reraising the exception. Then, write a `ReraiseHandler` that serves the sole purpose of calling `engine.terminate()`and reraising the exception. Register that as the last exception handler. This way, you can ensure that all exception handlers run, and use the last `ReraiseHandler` to choose which exceptions get raised and which get suppressed. 

I think it's ok to have an API like `engine.terminate(exception=e)` instead, but I think it's important to preserve the ability for the user to choose what exception is ultimately raised (so not exactly the one that triggered the handlers) as well as giving the option of not raising one at all.

**Comment by Nic-Ma** [comment]
Hi @amatsukawa ,

Thanks for your suggestion, it truly can work for the problem.
But if ignite can natively update this feature, it would be great.
Thanks.

**Comment by vfdev-5** [comment]
@amatsukawa thanks for the suggestion ! 

> If don't raise the exception again after handling, the training can't stop.

I think here we maybe have to fix current code where we catch `Exception` twice: 
1) whole run
https://github.com/pytorch/ignite/blob/f044948e4fe54d06d59299c2ecc823fc1e408790/ignite/engine/engine.py#L726
and 2) single epoch run
https://github.com/pytorch/ignite/blob/f044948e4fe54d06d59299c2ecc823fc1e408790/ignite/engine/engine.py#L793

Maybe, the first one (whole run, L726) will be enough to have a consistent behaviour and to stop the engine without calling `engine.terminate()`. 

If user raises an exception in a handler attached to `EXCEPTION_RAISED`, it will exit engine's code and no more handlers executed. This behaviour I think user would expect.

If user does not raise anything in all `EXCEPTION_RAISED` attached handlers, all handlers are executed, state is returned by `engine.run` and in some sense exception is silently catched... I'm not fan of that. So, the application can continue running after `engine.run`. Maybe, it would be better to automatically reraise the same exception at the end of all executed handlers.

>  I think it's important to preserve the ability for the user to choose what exception is ultimately raised (so not exactly the one that triggered the handlers) as well as giving the option of not raising one at all.

This can be nice. Maybe, not using an API like `engine.terminate(exception=e)` but somehow configuring the last reraised exception (from above).

**Comment by amatsukawa** [comment]
Thanks for the responses!

> If user does not raise anything in all EXCEPTION_RAISED attached handlers, all handlers are executed, state is returned by engine.run and in some sense exception is silently catched... I'm not fan of that. So, the application can continue running after engine.run. Maybe, it would be better to automatically reraise the same exception at the end of all executed handlers.

So would it no longer be possible to catch an exception and continue the engine execution? It seems important to preserve this case. Otherwise exceptions in the loop cannot be handled in a way that does not exit the `engine.run` function?

> This can be nice. Maybe, not using an API like engine.terminate(exception=e) but somehow configuring the last reraised exception (from above).

I think introducing some semantics like the engine will catch exceptions raised from handlers so all handlers can run, but will reraise the last exception that was by the handlers sounds good too.

**Comment by vfdev-5** [comment]
> So would it no longer be possible to catch an exception and continue the engine execution? It seems important to preserve this case. Otherwise exceptions in the loop cannot be handled in a way that does not exit the engine.run function?

I'm not sure if we intend to continue training if an exception was raised and `EXCEPTION_RAISED` was triggered. Currently, if an exception is encountered then the complete epoch (and not iteration) is then skipped.

**Comment by amatsukawa** [comment]
> I'm not sure if we intend to continue training if an exception was raised and EXCEPTION_RAISED was triggered. Currently, if an exception is encountered then the complete epoch (and not iteration) is then skipped.

Oh I didn't know that. I do crash on exceptions so it's fine, I'm not a fan of that :)

I agree that in common cases, we'd mostly like to raise an exception and stop training. But as you know from other tickets, my question is always "if I had a for-loop myself what could I do?" and I generally feel it's important to preserve a way to do anything I can do in such a loop in ignite. 

Maybe in a pinch while doing research, I want to throw an exception to short circuit skip a batch. I can zero gradients and reset everything in the exception block. Is it best engineering practice? Probably not. Bu sometimes it doesn't make sense to make that engineering investment if in 15 mins I might decide never mind, I don't want to skip batches after all. (I know there is a ticket to do this "properly" but my point is, I want to be able to do this in O(2 mins) not however long it takes to file a ticket and wait for a new feature.)

**Comment by vfdev-5** [comment]
> I agree that in common cases, we'd mostly like to raise an exception and stop training. But as you know from other tickets, my question is always "if I had a for-loop myself what could I do?" and I generally feel it's important to preserve a way to do anything I can do in such a loop in ignite.

Yes, I agree. How it would look like in a plain loop ?
```python
e = 0
try:  # first try-except
    while e < num_epochs:
        e += 1
        i = 0
        try: # second try-except
            while i < epoch_length:
                 try:   # this may be expensive, but user can do this inside its train_step()
                      train_step()
                 except Exception:
                      pass
            do_something()
        except Exception:
             pass
except Exception:
     pass
```
? 
Currently, the problem with engine's implementation is that we can not know where we are in "first try-except" or "second try-except"...


> (I know there is a ticket to do this example "properly" but my point is, I want to be able to do this in O(2 mins) not however long it takes to file a ticket and wait for a new feature.)

O(X) depends on our contributors :)

**Comment by alxlampe** [comment]
I wanted to create a new issue but just found this one 👍 
I was also expecting that an exception is raised in `Engine` even if we add event handlers for `EXCEPTION_RAISED`. 
In my use case, I also need to re-raise the exception after exception handling. 

> Currently, the problem with engine's implementation is that we can not know where we are in "first try-except" or "second try-except"...
If we have an exception in the "second try-except" and re-reraise it, we will end in the "first try-except" and do the exception handling again:

```python
from ignite.engine import Engine
from ignite.engine import Events


def _process_fun(engine, batch):
    if engine.state.iteration == 2:
        raise RuntimeError("error during iteration")

def _handle_exception(engine,exception):
    print("do something on exception")

def _reraise_exception(engine, exception):
    raise exception


if __name__ == '__main__':
    engine = Engine(_process_fun)
    engine.add_event_handler(Events.EXCEPTION_RAISED, _handle_exception)
    engine.add_event_handler(Events.EXCEPTION_RAISED, _reraise_exception)
    data = [1, 2, 3]
    engine.run(data, max_epochs=1)
```
Output:
```
Current run is terminating due to exception: error during iteration.
do something on exception 
Engine run is terminating due to exception: error during iteration.
do something on exception
[traceback]
```
Have you already noticed that?

> Maybe, the first one (whole run, L726) will be enough to have a consistent behavior and to stop the engine without calling engine.terminate().
This would of course solve the issue 👍 
Another solution would be to use an internal flag that indicates if the exception was handled in the inner loop. Don't know if this is a clean solution.

To pick up your discussion:   
IMO, always raising an exception is what most people would expect, even if you add exception handlers. 
And as @amatsukawa already said, there are good reasons, why you do not want to raise an exception in some cases. 

I think it is also important that there is no internal unexpected logic, if an exception is raised or not, as it is currently implement.  
There should be the option to always raise an exception (independent of the existence of exception handlers) or to not raise any exceptions.
What do you think?

**Comment by vfdev-5** [comment]
> There should be the option to always raise an exception (independent of the existence of exception handlers) or to not raise any exceptions.

@alxlampe yes, I agree that having both options could be nice. The question would be how to add this inside the current API in a simple and looking good way. If you have any suggestions, let's discuss that more.
